torch/csrc/distributed/c10d/FlightRecorder.hpp

```diff
@@ -188,25 +216,139 @@ struct FlightRecorder {
+        end,
+        c10::getTime(),
+        timeout_ms.count(),
+        isP2P,
+        std::nullopt,
+        std::nullopt,
+        std::nullopt,
+        {},
+        {},
+        {},
+        {},
+        {},
+        false};
+
+    for (const auto& input : inputs) {
+      c10::IntArrayRef sizes = input.sizes();
+      te.input_dtypes_.push_back(input.dtype().toScalarType());
+      te.input_dims_.push_back(static_cast<int64_t>(sizes.size()));
+      te.sizes_.insert(te.sizes_.end(), sizes.begin(), sizes.end());
+    }
+
+    for (const auto& output : outputs) {
+      c10::IntArrayRef sizes = output.sizes();
+      te.output_dtypes_.push_back(output.dtype().toScalarType());
+      te.output_dims_.push_back(static_cast<int64_t>(sizes.size()));
+      te.sizes_.insert(te.sizes_.end(), sizes.begin(), sizes.end());
+    }
+
+    if (entries_.size() < max_entries_) {
+      entries_.emplace_back(std::move(te));
+    } else {
+      entries_[next_++] = std::move(te);
+      if (next_ == max_entries_) {
+        next_ = 0;
+      }
+    }
+    return id_++;
+  };
 
   void record_pg_ranks(
       const std::tuple<std::string, std::string>& pg_name,
-      std::vector<uint64_t> ranks);
+      std::vector<uint64_t> ranks) {
+    if (!enabled_) {
+      return;
+    }
+    std::lock_guard<std::mutex> guard(mutex_);
+    pg_name_to_ranks_[pg_name] = std::move(ranks);
+  };
 
-  void record_accelerator_version(const std::string nccl_version);
+  void record_accelerator_version(const std::string nccl_version) {
+    if (!enabled_) {
+      return;
+    }
+    std::lock_guard<std::mutex> guard(mutex_);
+    nccl_version_ = std::move(nccl_version);
+  };
 
-  void update_state(Entry& r);
+  void update_state(Entry& r) {
+    if (r.start_ != nullptr) {
+      bool started = r.start_->query();
+      if (started && !r.time_discovered_started_) {
+        r.time_discovered_started_ = c10::getTime();
+      }
+    }
+    if (r.end_ != nullptr) {
+      bool completed = r.end_->query();
+      if (completed && !r.time_discovered_completed_) {
+        r.time_discovered_completed_ = c10::getTime();
+      }
+    }
+  };
 
-  std::vector<Entry> dump_entries();
+  std::vector<Entry> dump_entries() {
+    std::lock_guard<std::mutex> guard(mutex_);
+    std::vector<Entry> result;
+    result.reserve(entries_.size());
+    result.insert(
+        result.end(),
+        entries_.begin() + static_cast<std::ptrdiff_t>(next_),
+        entries_.end());
+    result.insert(
+        result.end(),
+        entries_.begin(),
+        entries_.begin() + static_cast<std::ptrdiff_t>(next_));
+    // query any remaining events
+    for (auto& r : result) {
+      update_state(r);
+      r.start_ = r.end_ = nullptr;
+    }
+    return result;
+  };
 
   // Returns the entry with the given id, if it exists. Otherwise, returns
   // std::nullopt.
-  std::optional<Entry> getEntry(std::optional<size_t> id);
+  std::optional<Entry> getEntry(std::optional<size_t> id) {
```

Comment:
nit: better not to move all the code to .hpp just becuase of templating? I think you can keep functions like this in the .cpp,

and even for other functions i wonder if its worth making our own 'FlightEvent' wrapper that is templated over different event types and then just leave all the FR logic in terms of FlightEvent?

open to other ideas too

Reply:
Keeping even functions which are not dependent on Event will give link errors.

Reply:
yeah, unfortunately since it's all templatized we need it to all be in the header file.

We could potentially split this into two copies, one with just the high level API and then a second header file with "FlightRecorderImpl" to try to minimize the amount of headers included in all files. That way we only depend on the impl form the PG directly but given this is a fairly specialized interface we may be able to get that isolation regardless

For these big header-only libs I usually use folly as a prior art for how to structure. I.e. the Future class https://www.internalfb.com/code/fbsource/[6d0f557f832a72e3a7c6323a03123fab17993171]/fbcode/folly/futures/

Reply:
Somehow with the help of Ke, we can do something like this, I updated the PR and you guys can see if it makes sense.